pytensor/tensor/shape_ops.py

```diff
@@ -0,0 +1,461 @@
+from collections.abc import Iterable, Sequence
+from itertools import pairwise
+
+import numpy as np
+from numpy.lib._array_utils_impl import normalize_axis_tuple
+
+from pytensor import Variable
+from pytensor.graph import Apply
+from pytensor.graph.op import Op
+from pytensor.tensor import TensorLike, as_tensor_variable
+from pytensor.tensor.basic import expand_dims, join, split
+from pytensor.tensor.math import prod
+from pytensor.tensor.shape import ShapeValueType
+from pytensor.tensor.type import tensor
+from pytensor.tensor.variable import TensorConstant, TensorVariable
+
+
+class JoinDims(Op):
+    def __init__(self, axis: Sequence[int] | int | None = None):
+        self.axis = axis
```

Comment:
refuse negative axis at the op level (to avoid redundant parametrizations) and raise if non contiguous.